### ISSUE TITLE
build on windows

### ISSUE THREAD

**Issue by marcelveldt** [michaelherger/spotty#1]
Great work on this!

I was looking for this exact same thing to replace libspotify in my Kodi addon and only need the raw audio pipe. If I may ask, how did you manage to build this on windows ?

Got it working on my mac within a few minutes but windows keeps complaining about mdns and portaudio.

**Comment by michaelherger** [comment]
Oh, true, I still manually disable (comment out) mdns in librespot's Cargo.toml. That's why I point to the neighboring folder in spotty's Cargo.toml in that case. I still haven't figured out how I can conditionally enable/disable includes in Cargo.toml.

- clone librespot from my repository to a folder next to the spotty clone
- checkout the "spotty" branch in that librespot clone
- modify Cargo.toml in the spotty clone folder to use that clone rather then directly from GitHub:
`[dependencies.librespot]`
`#git = "https://github.com/michaelherger/librespot.git"`
`#branch = "change-notification"`
`path = "../librespot"`
`default-features = false`
- run Cargo from the spotty folder

portaudio should have been taken care off by the default-features = false statement in Cargo.toml.

**Comment by michaelherger** [comment]
FWIW: I just pushed a change to my librespot repository which automatically deals with the mdns issue on Windows. You should be able to compile without any manual change.

**Comment by marcelveldt** [comment]
Wow, fantastic. Thanks for your help. Will try later today once finished work.
Do you have any plans of integrating this into a new Spotify app for the LMS server ?

I just recently discovered LMS and must say I really like it.
Looking into the Spotify implementation I discovered that the helper application is also still using lib spotify, which will be abandoned by Spotify this year, so same issue as I'm having with my Kodi application for Spotify.

Very recent, Spotify has updated their webapi with spotify connect possibilities so you'll be able to control connect devices. I discovered that every player of Spotify, even the web player, is now controlled by the connect protocol so that seems to be the new direction they're heading. 

That got me thinking to use that same approach for the LMS plugin. The helper application which is responsible for playback streaming to devices and representing them as connect devices to spotify.

I'm actually currently looking in to the Shairport implementation in LMS to see if that same approach can be used to get spotify connect going. In that case the spotify helper broadcasts all LMS players as connect devices and is also available to react to spotify playback requests directly. As I have no experience with perl this is not going very fast but I'm making some progress.

**Comment by michaelherger** [comment]
http://forums.slimdevices.com/showthread.php?t=107418 - it doesn't offer the Spotify Connect feature, but only replaces the player backend for LMS. Very much like my previous Spotify Protocol Handler plugin.

**Comment by marcelveldt** [comment]
Oh wow, you even went that far of creating exactly that :-)
Thanks, I will give it a test shot. 

BTW: Just have built the updated source and now it compiles perfectly fine. Playback however isn't working when piping to VLC but that could also be my own stupidity. Will test with your version first.

Out of curiosity:

- Do you have any idea if this will work with multiple spotify accounts ? So I can have multiple spotty streams throughout the house ?

- What about my idea to also use the connect feature to present each player as a connect device ? A bit the same as how sharper currently works where your squeeze players show up as airplay targets. Os is that perhaps a bridge too far ?

Anyways, thanks a lot for this.

**Comment by michaelherger** [comment]
- You can have multiple streams by launching multiple copies of spotty. As it's not running as a daemon, but rather a helper to pull a track and pipe it elsewhere, it only lives for the time it takes to get a track. Yes, there's some overhead to this, but it works very well on a Pi3 (< 10% CPU load while playing one stream and transcoding to Flac through LMS)
- Connect in LMS comes with a bunch of UX issues. It might work fine if you use one of the playback engines. But in piping mode not so much. I think there's a bug in librespot, too, which causes skipping to the next track as soon as a track has been fully downloaded - which can be within seconds. See https://github.com/plietar/librespot/issues/175.
- VLC might need some parameters telling it what bitrate etc. I needed to do the same for mp3/flac transcoding: --channels=2 --sample-rate=44100 --bps=16 --endian=little --sign=signed. Don't know VLC enough to tell you how you would define them for it. But this might give you a start.

**Comment by marcelveldt** [comment]
this is absolutely fantastic, the skip after a few seconds was exactly what I was experiencing in because I was thinking it would get a buffer from spotify streaming the audio realtime but instead it got the whole track, really didn't think about that!

Thanks for your work on this. I'll see if I can replace my lib spotify implementation in Kodi spotify addon with your spotty one.

Not that I use that plugin a lot myself now that I discovered LMS and squeezelite ;-)

I will give your spotty LMS plugin a try later today.
So basically what you're saying is that I can run multiple instances of the plugin, just give them a different ID. But they will all hijack the same spotify:// protocol.

**Comment by marcelveldt** [comment]
just to confirm: working perfectly fine now with VLC !

**Comment by marcelveldt** [comment]
Testing the LMS plugin without any luck. Playback stops immediately. LMS installed on Synology NAS, latest 7.9.0 version. Do I need to switch to dev 7.9.1 for this ?

Please say so if you want me to move these questions to the forums as these aren't related to my original question anymore.

**Comment by marcelveldt** [comment]
Installed a fresh copy of LMS 7.9.1 and still no dice. Looks like the spotty binary is crashing badly.
I do not see any logging in the server log at all from spotty plugin.
I do however see this message in my system log:

[39959.964415] traps: spotty[14424] trap invalid opcode ip:7f6e5e691e61 sp:7f6e5cdf7880 error:0 in spotty[7f6e5e3d1000+46f000]

I have the feeling this has something to do with the fact I'm running this on a Synology. It is a x86 model however.

**Comment by marcelveldt** [comment]
OK, swapped the i686 binary with the x86_64 version and now it's playing ;-)
So, looks like the issue is with the x86_64 version of the spotty binary

I'll try to build it from source in a VM to see if I can spot the issue

**Comment by michaelherger** [comment]
Could you please continue LMS specifics in that thread I linked? And if you do so, please include information about what LMS reports on Settings/Information about the platform, perl version, and path for helper apps. Thanks!

**Comment by michaelherger** [comment]
What did you have to do to get it working with VLC?

**Comment by marcelveldt** [comment]
Will do, no problem.

For VLC, I used this: spotty.exe with params | vlc.exe - --demux=rawaud --rawaud-channels 2 --rawaud-samplerate 44100

**Comment by marcelveldt** [comment]
Hi Michael, it appears that librespot is no longer working, can you confirm ?
https://github.com/plietar/librespot/issues/190

**Comment by michaelherger** [comment]
A scary development :-(

**Comment by marcelveldt** [comment]
Yes, indeed. Well it appears that it can be fixed with a simple change but the question is if Spotify is now blocking intentionally

**Comment by marcelveldt** [comment]
BTW: This is the resolution for the issue with the x64 binary on some intel NAS platforms I told you about on the forums:

https://github.com/marcelveldt/librespot/commit/f792da18a4345a17065e51e9e6a69c7d136d3722

Just use a rust-crypto version with the patch already supplied to build librespot and you're al set.
